src/app/clusters/time-synchronization-server/time-synchronization-server.h

```diff
@@ -0,0 +1,170 @@
+
+#include <app-common/zap-generated/cluster-objects.h>
+#include <lib/support/Span.h>
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace TimeSynchronization {
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::TimeSynchronization;
+using namespace chip::app::Clusters::TimeSynchronization::Attributes;
+using chip::TimeSyncDataProvider;
+using chip::Protocols::InteractionModel::Status;
+
+class TimeSynchronizationServer
+{
+    struct timeZoneName
+    {
+        char name[64];
+    };
+
+public:
+    void Init();
+
+    static TimeSynchronizationServer & Instance(void);
+
+    TimeSyncDataProvider GetDataProvider(void) { return mTimeSyncDataProvider; }
+
+    CHIP_ERROR SetTrustedTimeSource(DataModel::Nullable<TimeSynchronization::Structs::TrustedTimeSourceStruct::Type> tts)
+    {
+        if (!tts.IsNull())
+        {
+            mTrustedTimeSource.SetNonNull(tts.Value());
+            mTimeSyncDataProvider.StoreTrustedTimeSource(tts.Value());
+        }
+        else
+        {
+            mTrustedTimeSource.SetNull();
+            mTimeSyncDataProvider.ClearTrustedTimeSource();
+        }
+
+        return CHIP_NO_ERROR;
+    }
+
+    CHIP_ERROR SetDefaultNtp(DataModel::Nullable<chip::MutableByteSpan> dntp)
+    {
+        if (!dntp.IsNull())
+        {
+            memcpy(mDefaultNtpBuf, dntp.Value().data(), dntp.Value().size());
+            mDefaultNtp.SetNonNull(MutableByteSpan(mDefaultNtpBuf, dntp.Value().size()));
+            return mTimeSyncDataProvider.StoreDefaultNtp(mDefaultNtp.Value());
+        }
+        else
+        {
+            mDefaultNtp.SetNull();
+            return mTimeSyncDataProvider.ClearDefaultNtp();
+        }
+        return CHIP_NO_ERROR;
+    }
+
+    CHIP_ERROR SetTimeZone(DataModel::DecodableList<TimeSynchronization::Structs::TimeZoneStruct::Type> tz)
+    {
+        auto mTzL = mTimeZoneList.begin();
+        auto tzL  = tz.begin();
+        size_t i  = 0;
+
+        while (tzL.Next())
+        {
+            mTzL[i].offset  = tzL.GetValue().offset;
+            mTzL[i].validAt = tzL.GetValue().validAt;
+            if (tzL.GetValue().name.HasValue())
+            {
+                const char * buf = tzL.GetValue().name.Value().data();
+                size_t len       = tzL.GetValue().name.Value().size();
+                Platform::CopyString(mNames[i].name, chip::CharSpan(buf, len));
+            }
+            i++;
+        }
+
+        return mTimeSyncDataProvider.StoreTimeZone(mTimeZoneList);
+    }
+
+    CHIP_ERROR SetDSTOffset(DataModel::DecodableList<TimeSynchronization::Structs::DSTOffsetStruct::Type> dst)
+    {
+        auto mDstL = mDstOffList.begin();
+        auto dstL  = dst.begin();
+        size_t i   = 0;
+
+        while (dstL.Next())
+        {
+            mDstL[i] = dstL.GetValue();
+            i++;
+        }
+
+        return mTimeSyncDataProvider.StoreDSTOffset(mDstOffList);
```

Comment:
Hold on.  How are we using the size of the incoming list?  We don't seem to be....